File: internal/elk/cross_check_test.go
// cross_check_test.go は Go `internal/elk` とフロント `frontend/src/layout/elk.ts`
// の ELK 入力グラフ「構造」の一致を担保するクロスチェックテスト（ADR-0001）。
//
// レイアウト計算そのものはブラウザ側（elkjs）が担うため一致対象にはしない。
// ここで縛るのは「一致すべき構造」= (1) primary グループ所属（groupNode 階層）と
// (2) エッジ集合（親→子、sanitizeID 正規化済み）である。
//
// 共有 fixture（`frontend/src/layout/testdata/elk-cross-check-fixtures.json`）を
// Go の `*model.Schema` にデコードし、buildRoot から構造サマリを抽出して期待値
// （`frontend/src/layout/testdata/expected/<name>.elk-structure.json`）と比較する。
// 期待値は Go 側を「真実の単一の源」とし、UPDATE_GOLDEN=1 で再生成できる。
// TS 側の cross-check.test.ts は同じ期待値ファイルへの一致を検証する。
package elk

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"testing"

	"github.com/unok/erdm/internal/model"
	"github.com/unok/erdm/internal/testutil/golden"
)

const (
	elkFixturePath = "../../frontend/src/layout/testdata/elk-cross-check-fixtures.json"
	elkExpectedDir = "../../frontend/src/layout/testdata/expected"
)

// elkStructureGroup は 1 つの primary グループとその所属テーブル ID 列。
type elkStructureGroup struct {
	ID     string   `json:"id"`
	Tables []string `json:"tables"`
}

// elkStructureEdge は 1 本の FK エッジ（親→子）。
type elkStructureEdge struct {
	Source string `json:"source"`
	Target string `json:"target"`
}

// elkStructureSummary は ELK 入力グラフの「一致すべき構造」だけを取り出したもの。
// 座標・サイズ・layoutOptions は Go/TS で正当に異なるため含めない。
type elkStructureSummary struct {
	Groups    []elkStructureGroup `json:"groups"`
	Ungrouped []string            `json:"ungrouped"`
	Edges     []elkStructureEdge  `json:"edges"`
}

// summarizeRoot は elkRoot から構造サマリを抽出する。TS 側の同名処理と一致させる。
func summarizeRoot(root *elkRoot) elkStructureSummary {
	summary := elkStructureSummary{
		Groups:    []elkStructureGroup{},
		Ungrouped: []string{},
		Edges:     []elkStructureEdge{},
	}
	for _, child := range root.Children {
		if len(child.Children) > 0 {
			tables := make([]string, 0, len(child.Children))
			for _, member := range child.Children {
				tables = append(tables, member.ID)
			}
			summary.Groups = append(summary.Groups, elkStructureGroup{ID: child.ID, Tables: tables})
			continue
		}
		summary.Ungrouped = append(summary.Ungrouped, child.ID)
	}
	for _, e := range root.Edges {
		summary.Edges = append(summary.Edges, elkStructureEdge{Source: e.Sources[0], Target: e.Targets[0]})
	}
	return summary
}

func TestElkStructure_CrossCheckWithFrontend(t *testing.T) {
	raw, err := os.ReadFile(elkFixturePath)
	if err != nil {
		t.Fatalf("read fixtures: %v", err)
	}
	var fixtures map[string]*model.Schema
	if err := json.Unmarshal(raw, &fixtures); err != nil {
		t.Fatalf("unmarshal fixtures: %v", err)
	}

	names := make([]string, 0, len(fixtures))
	for name := range fixtures {
		names = append(names, name)
	}
	sort.Strings(names)

	for _, name := range names {
		t.Run(name, func(t *testing.T) {
			schema := fixtures[name]
			if schema == nil {
				t.Fatalf("fixture %s missing", name)
			}
			summary := summarizeRoot(buildRoot(schema))
			got, err := json.MarshalIndent(summary, "", "  ")
			if err != nil {
				t.Fatalf("marshal summary: %v", err)
			}
			got = append(got, '\n')
			golden.Compare(t, got, filepath.Join(elkExpectedDir, name+".elk-structure.json"))
		})
	}
}

File: frontend/src/layout/cross-check.test.ts
// クライアント側 `buildElkInput` と Go 側 `internal/elk` の ELK 入力グラフ
// 「構造」の一致を検証するクロスチェックテスト（ADR-0001）。
//
// 共有 fixture（`testdata/elk-cross-check-fixtures.json`）から ELK 入力を構築し、
// 「一致すべき構造」= primary グループ所属 + エッジ集合 を抽出して、Go 側が生成した
// 期待値（`testdata/expected/<name>.elk-structure.json`）と比較する。
// 期待値は Go 側 `internal/elk/cross_check_test.go` が真実の単一の源として生成する。
// 期待値が古くなった場合は Go 側テストが先に失敗する。

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import type { ElkNode } from 'elkjs/lib/elk.bundled.js'
import { describe, expect, it } from 'vitest'
import type { Schema } from '../model'
import { buildElkInput } from './elk'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const FIXTURES_PATH = path.join(__dirname, 'testdata', 'elk-cross-check-fixtures.json')
const EXPECTED_DIR = path.join(__dirname, 'testdata', 'expected')

interface StructureSummary {
  groups: { id: string; tables: string[] }[]
  ungrouped: string[]
  edges: { source: string; target: string }[]
}

// summarize は ELK 入力ルートから「一致すべき構造」を抽出する。
// Go 側 `summarizeRoot` と同じ規則: children を持つノード=groupNode、
// それ以外=ungrouped テーブル。エッジは親→子（sources[0]→targets[0]）。
function summarize(root: ElkNode): StructureSummary {
  const groups: { id: string; tables: string[] }[] = []
  const ungrouped: string[] = []
  for (const child of root.children ?? []) {
    if (child.children && child.children.length > 0) {
      groups.push({ id: child.id, tables: child.children.map((m) => m.id) })
    } else {
      ungrouped.push(child.id)
    }
  }
  const edges = (root.edges ?? []).map((e) => ({
    source: e.sources[0]!,
    target: e.targets[0]!,
  }))
  return { groups, ungrouped, edges }
}

function loadFixtures(): Record<string, Schema> {
  return JSON.parse(fs.readFileSync(FIXTURES_PATH, 'utf8')) as Record<string, Schema>
}

function loadExpected(name: string): StructureSummary {
  const raw = fs.readFileSync(path.join(EXPECTED_DIR, `${name}.elk-structure.json`), 'utf8')
  return JSON.parse(raw) as StructureSummary
}

describe('ELK structure cross-check vs Go reference', () => {
  const fixtures = loadFixtures()
  const names = Object.keys(fixtures).sort()

  it('covers every fixture with an expected golden', () => {
    expect(names.length).toBeGreaterThan(0)
  })

  for (const name of names) {
    it(`matches Go structure for ${name}`, () => {
      const got = summarize(buildElkInput(fixtures[name]!))
      expect(got).toEqual(loadExpected(name))
    })
  }
})
